clustermesh-apiserver/etcdinit/root.go

```diff
@@ -0,0 +1,122 @@
+	einit "github.com/cilium/cilium/pkg/etcd/init"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/spf13/cobra"
+	client "go.etcd.io/etcd/client/v3"
+)
+
+var (
+	log                     = logging.DefaultLogger.WithField(logfields.LogSubsys, "etcdinit")
+	etcdBinaryLocation      = "/usr/bin/etcd"
+	etcdDataDir             = ""
+	etcdClusterName         = ""
+	etcdInitialClusterToken = ""
+	debug                   = false
+)
+
+func NewCmd() *cobra.Command {
+	rootCmd := &cobra.Command{
+		Use:   "etcdinit",
+		Short: "Initialise an etcd data directory for use by the etcd sidecar of clustermesh-apiserver",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return InitEtcdLocal()
+		},
+	}
+	rootCmd.Flags().StringVar(&etcdDataDir, "data-dir", defaults.EtcdDataDir, "etcd data directory")
+	rootCmd.Flags().StringVar(&etcdClusterName, "cluster-name", defaults.EtcdClusterName, "etcd cluster name")
+	rootCmd.Flags().StringVar(&etcdInitialClusterToken, "initial-cluster-token", defaults.EtcdInitialClusterToken, "etcd initial cluster token")
+	rootCmd.Flags().BoolVar(&debug, "debug", false, "Debug log output")
+	return rootCmd
+}
+
+func InitEtcdLocal() error {
+	ctx, cancelFn := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancelFn()
+
+	if debug {
+		logging.SetLogLevelToDebug()
+		log.Debug("Debug logging enabled")
+	}
+
+	log := log.
+		WithField("etcdBinaryLocation", etcdBinaryLocation).
+		WithField("etcdDataDir", etcdDataDir).
+		WithField("etcdClusterName", etcdClusterName).
+		WithField("etcdInitialClusterToken", etcdInitialClusterToken)
+
+	// When the clustermesh-apiserver is launched, we create a new etcd. We don't support persistence of any kind
+	// here. So we just delete the contents of the data directory. It should be empty, but if the initialization failed
+	// we might re-run this as an init container.
+	log.Info("Deleting contents of data directory")
+	dir, err := os.ReadDir(etcdDataDir)
+	if err != nil {
+		return err
+	}
+	for _, d := range dir {
+		log.WithField("path", d.Name()).Debug("Removing file/directory in data dir")
+		os.RemoveAll(path.Join(etcdDataDir, d.Name()))
+	}
+
+	log.Info("Starting localhost-only etcd process")
+	// Specify the full path to the etcd binary to avoid any PATH search binary replacement nonsense
+	endpoint := "http://127.0.0.1:2379"
+	etcdCmd := exec.CommandContext(ctx, etcdBinaryLocation,
+		fmt.Sprintf("--data-dir=%s", etcdDataDir),
+		fmt.Sprintf("--name=%s", etcdClusterName),
+		// This probably won't work on an IPv6 only cluster
+		fmt.Sprintf("--listen-client-urls=%s", endpoint),
+		fmt.Sprintf("--advertise-client-urls=%s", endpoint),
+		fmt.Sprintf("--initial-cluster-token=%s", etcdInitialClusterToken),
+		"--initial-cluster-state=new",
+		"--auto-compaction-retention=1")
+
+	// Exec the etcd binary, which ultimately calls fork(2) under the hood. We're attached to STDIN and STDOUT for the
```

Comment:
Is there any stderr/stdout output we'd need to capture from the etcd binary prior to it forking?

Reply:
I did consider capturing stdout/stderr, but I wasn't sure what to do with it. I could log it out too, but then it might get confusing, especially if you're using structured logging and the logging format is different between the two. Or I could "wrap" each line as a string in a log message from this package. Not sure. Did you have any thoughts?

Reply:
So, I tired this. Unless I'm missing something, it seems very difficult to capture stdout and stderr from a forked process without go routines that are very likely to cause locks. There is [go-cmd](https://github.com/go-cmd/cmd) that is designed for this purpose, but I didn't want to add an extra dependency to Cilium if I could help it. But if we want the output captured for debug I can add it and do that.

Reply:
Yeah capturing the forked logs seems not worth the effort, was mostly wondering for the logs prior to the fork (i.e. say the fork syscall failed and etcd bailed out).

Reply:
I don't think there's a provision for that in the cmd library.

If you follow the code path for `cmd.Start` (in go1.21.1) you eventually end up in the function [`forkAndExecInChild`](https://github.com/golang/go/blob/go1.21.1/src/syscall/exec_linux.go#L127). That function actually does [clone(2)](https://man7.org/linux/man-pages/man2/clone.2.html) on Linux [here](https://github.com/golang/go/blob/go1.21.1/src/syscall/exec_linux.go#L311), returns from the parent, and continues the rest of the function in the child process. In the child process it then calls [execve(2)](https://man7.org/linux/man-pages/man2/execve.2.html) [here](https://github.com/golang/go/blob/go1.21.1/src/syscall/exec_linux.go#L631) to move into the called process (etcd).

I don't think there's any opportunity to capture logs before the fork (technically clone on Linux, but same concept), because the fork happens before etcd is executed.